### PR TITLE
test(protocol-designer): e2e batch edit automation cases

### PR DESCRIPTION
**Pull request by reachojhaneha**
# Overview

Writing the e2e tests for batch

# Changelog

- Automation cases for batch edit

# Risk assessment
low

Comment
work in progress, this PR needs feature flag to be removed. Blocked by #7446

### CODE REVIEW

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+  // import the batchEdit.json to PD
+  it('Verify Flowrate, dublicate, delete functionality in Batch Edit Mode', () => {
+    importProtocol()
+    openDesignTab()
+
+    // enter into the batch edit mode
+    cy.get('[data-test="StepItem_1"]').click({
+      ctrlKey: true,
+    })
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: after this can we guarantee that the banner shows up or something? just to show we are in batch edit mode

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Range selection with shift
+    cy.get('[data-test="StepItem_3"]').click({
+      shiftKey: true,
+    })
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js:24]: can we validate that the steps actually get selected? maybe via styles or something, we can look into this

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Change the Flowrate to 100 and Save
+    cy.get('[name="aspirate_flowRate"]').click()
+    cy.get('input[name="aspirate_flowRate_customFlowRate"]').type('100')
+    cy.get('button').contains('Done').click()
+    cy.get('button').contains('save').click()
+
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js:35]: can we validate that he flow rate gets saved? maybe by reopening the form and making sure the value has changed

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    // Expand ALL steps
+    cy.get('#ClickableIcon_expand').click()
+
+    // Delete the dublicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
+
+    // Exit batch edit mode
+    cy.get('button').contains('exit batch edit').click()
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: Same thing with all of these, we should validate that the steps have been deleted/duplicated/expanded. And that we've successfully exited batch edit mode

**Review by shlokamin — CHANGES_REQUESTED** [review]
Prior to release we should make sure we assert that each of the actions we're doing were successful. I'm down to pitch in here!

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 1 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: Good Point. I am doing set of actions in this script but not verifying them. Will verify these steps.

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 2 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js:24]: will do

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 3 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js:35]: will do

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 4 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: will do

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 1 above)
Comment (shlokamin) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: cool, lmk if you'd like any support. and thanks for this, we really appreciate it!

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    // Verify that transfer step 1 has Flowrate value 100
+    cy.get('[data-test="StepItem_1"]').click({
+      shiftKey: true,
+    })
+    cy.get('input[name="aspirate_flowRate"]').should('have.value', 100)
+
+    // Range selection with shift
+    cy.get('[data-test="StepItem_3"]').click({
+      shiftKey: true,
+    })
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: how come this range selection is needed? it looks like just after there is another range selection on `StepItem_4`

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    // Dublicate the selected steps
+    cy.get('#ClickableIcon_duplicate').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('9 steps selected')
+      .should('exist')
+
+    // Delete the dublicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: this doesn't quite validate that a duplicate and delete went through. could we also verify that when you click `duplicate` there will be a total of 18 steps? could just be: 

```js
cy.get('[data-test="StepItem_18"]').should('exist') //there are now 18 step items
```

and then after the `delete`:

```js
cy.get('[data-test="StepItem_10"]').should.not.('exist') // there are now only 9 stepitems
```

or something like that... 

does that make sense?

Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: this is addressed, not after duplicating i verify that 18 steps are there and deleting steps verify the remaining steps too.

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 5 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: removed this step

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 7]
```diff
@@ -71,13 +66,21 @@ describe('Batch Edit Transform', () => {
...
     cy.get('#StepSelectionBannerComponent_numberStepsSelected')
-      .contains('9 steps selected')
+      .contains('18 steps selected')
       .should('exist')
 
+    cy.get('[data-test="StepItem_9"]').click({
+      shiftKey: true,
+    })
+
     // Delete the dublicated steps
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: `dublicated  => duplicated` 😄

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
+describe('Batch Edit Transform', () => {
+  beforeEach(() => {
+    cy.visit('/')
+    cy.closeAnnouncementModal()
+  })
+
+  // import the batchEdit.json to PD
+  it('Verify Flowrate, dublicate, delete functionality in Batch Edit Mode', () => {
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: `dublicate  => duplicate` 😄

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+      .contains('A1')
+      .should('exist')
+
+    // Select all the steps
+    cy.get('#ClickableIcon_select').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('9 steps selected')
+      .should('exist')
+
+    // Dublicate the selected steps
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js]: `dublicate => duplicate 😄`

**Review by shlokamin — COMMENTED** [review]
Some small typos, but it also looks like CI is failing. Looks this test cannot find the `exit batch edit mode` button

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 7 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: Test is not finding exit batch edit mode button, as its not able to enter the batch edit form. Does the E2E run on production environment? As this case passes on sandbox

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 9 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: Thanks, will correct this

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 7 above)
Comment (shlokamin) [reply protocol-designer/cypress/integration/batchEdit.spec.js]: no they only run on sandbox

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    cy.get('[data-test="StepItem_9"]').click({
+      shiftKey: true,
+    })
+
+    // Delete the duplicated steps
+    cy.get('#ClickableIcon_delete').click()
+    cy.get('button').contains('delete steps').click()
+    cy.get('#StepSelectionBannerComponent_numberStepsSelected')
+      .contains('1 steps selected')
+      .should('exist')
```
Comment (shlokamin) [inline protocol-designer/cypress/integration/batchEdit.spec.js:82]: This part I do not follow...

1. We duplicated our 9 steps, so now we have 18 steps selected
2. We shift + click on step number 9 (at this point we would have deselected 9 steps, so 9 should now be selected right?)
3. Delete the selected steps (which theoretically is 9 steps)
4. Then only 1 item is selected

How is it that only 1 item is left selected? Is this because after we delete, we are always only left with one selected item?

Comment (shlokamin) [reply protocol-designer/cypress/integration/batchEdit.spec.js:82]: i think i answered my own q above...

would it be worth checking that we only have 9 steps left after clicking `delete`?

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+    cy.get('[id=Text_noSharedSettings]').contains(
+      'Batch editing of settings is only available for Transfers',
+      {
+        matchCase: false,
+      }
+    )
+
+    // Expand ALL steps
```
Comment (IanLondon) [inline protocol-designer/cypress/integration/batchEdit.spec.js:54]: I think this comment is inaccurate, it's really expanding just the selected steps (1 and 4 at this point)

File: protocol-designer/cypress/integration/batchEdit.spec.js [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    cy.get('#ClickableIcon_expand').click()
+    cy.get('[data-test="SubstepRow_aspirateWell"]')
```
Comment (IanLondon) [inline protocol-designer/cypress/integration/batchEdit.spec.js:56]: This is testing that the step is expanded, right? Maybe we should make sure specifically that steps 1 & 4 are expanded and that eg step 2 isn't? Right now, this would pass if any one step was expanded, I think.

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 10 above)
Comment (IanLondon) [reply protocol-designer/cypress/integration/batchEdit.spec.js:82]: Might be nice to have a comment just pointing out that after the shift-click, Steps 1 to 9 should be selected (bc it's hard to follow in my head as a read it)

Also agree I think we might want to delete, then select all, and confirm that 9 steps are remaining (instead of, or in addition to, checking that 1 step is currently selected immediately after clicking delete)

**Review by IanLondon — CHANGES_REQUESTED** [review]
Looks great! Just have a few comments.

Also, could you merge the latest edge on to this? It has some changes that have already been merged into edge (eg StepSelectionBanner + SubstepRow) and also since this branch may be getting a little older, it'd be good to update it so that we're not surprised by something unexpected happening when it is merged

File: protocol-designer/cypress/integration/batchEdit.spec.js (hunk 10 above)
Comment (reachojhaneha) [reply protocol-designer/cypress/integration/batchEdit.spec.js:82]: this is what i tried: selected all the steps and shift + step 9 to deselect steps 9 to 18. Delete steps 1-8 steps as they are selected. Once all the selected steps are deleted, next step becomes selected as for batch edit one step is always selected(imo this is a verification too).
Now, about selecting steps 9-18 steps, select-all(id - #ClickableIcon_select)  is hidden by another element <div class="lists__title_bar__3n-Et lists__clickable__2o_y_">...</div>, and cypress is not able to access it. Hence i took the other route mentioned above.

**Comment by shlokamin** [comment]
Looks like e2e cases are failing: 

```
AssertionError: Timed out retrying after 4000ms: Expected to find content: 'Batch editing of settings is only available for Transfers' within the element: <p#Text_noSharedSettings.Text-sc-1wb1h0f-0.nVDCX> but never did.
```

**Comment by Kadee80** [comment]
> Looks like e2e cases are failing:
> 
> ```
> AssertionError: Timed out retrying after 4000ms: Expected to find content: 'Batch editing of settings is only available for Transfers' within the element: <p#Text_noSharedSettings.Text-sc-1wb1h0f-0.nVDCX> but never did.
> ```

Might this be because we changed the text to Transfers or Mix?

**Comment by shlokamin** [comment]
Ahhh yep, we have to change the cypress query to `"Batch editing of settings is only available for Transfer or Mix steps"`

**Review by IanLondon — APPROVED** [review]
🎉

**Review by shlokamin — APPROVED** [review]
